Add internal assert for tangent layout mismatch for view ops

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68629
* #68628
* #67045
* #67041
* #67820
* #66294
* __->__ #66293
* #66292

This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

soulitzer committed:
Add internal assert for tangent layout mismatch for view ops

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add internal assert for tangent layout mismatch for view ops"




[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`. 

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset 

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset 

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -165,13 +165,18 @@ void AutogradMeta::set_fw_grad(const at::TensorBase& new_grad_base, const at::Te
 
     // Enforce the basic layout constraint
     if (!has_same_meta(new_grad, self)) {
+      if (is_view_ && !is_make_dual) {
```
I still don't think we should special case make_dual here. It makes the code more complex for no benefit.

soulitzer replied:
Maybe we should update codegen for InplaceOrView to have `is_fw_differentiable` as `false` when the input tensor does not have a forward grad? That would also fix the issue.

albanD commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -29,6 +29,7 @@ BatchedTensorImpl::BatchedTensorImpl(Tensor value, BatchDims bdims)
     sizes_and_strides_.size_at_unchecked(dim) = value_sizes.at(actual_dim);
     sizes_and_strides_.stride_at_unchecked(dim) = value_strides.at(actual_dim);
   }
+  storage_offset_ = value_.storage_offset();
```
Why is this the right value to use? Should we remove the potential offset from the batch dims?

soulitzer replied:
It is okay for each slice of the batched tensor to inherit the same offset, if that is what you mean (for example consider an op like slice[1:]). I don't think the potential offset from the batch dims matters in this particular case because these properties are for the logical tensor.

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -576,6 +577,18 @@ Tensor _new_zeros_with_same_meta_batching_rule(
+  DispatchKeySet after_batched_keyset =
+      DispatchKeySet(DispatchKeySet::FULL_AFTER, c10::DispatchKey::Batched);
+  return at::redispatch::_make_dual(ks & after_batched_keyset, primal, tangent, level);
+}
```
What I wanted to do here is to basically say that _make_dual is composite wrt the Batched key, so that we can redispatch to the autograd key first. Maybe there is a better way @zou3519?

albanD replied:
Can you register a make fallthrough to do that?

zou3519 replied:
I'm confused: why doesn't `m.impl("_make_dual", native::_make_dual);` work?

soulitzer replied:
If we had `native::_make_dual`, it would go directly to our implementation (which all it does is call `primal.view(primal.shape())`. We'd like for it to go through our kernel in VariableTypeManual, which does other necessary stuff.

soulitzer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -98,8 +98,10 @@
     CompositeExplicitAutograd: _fw_primal
 
-- func: _make_dual(Tensor(a) primal, Tensor tangent, int level) -> Tensor(a)
+- func: _make_dual(Tensor(a) self, Tensor tangent, int level) -> Tensor(a)
```
Currently functionalization pass codegen seems to assume that first input for view functions must be `self`. Is this intended @bdhirsh?

albanD replied:
We do have views that don't have self as arg: https://github.com/pytorch/pytorch/blob/6b1d8e5bb20fc56d937f3177f0c3cc6d0d37e0bd/tools/autograd/gen_inplace_or_view_type.py#L71 ?

zou3519 replied:
Probably not intended? Could you file an issue please and cc Brian on it?

bdhirsh replied:
Yep, the functionalization codegen makes some assumptions about the schema of views. It doesn't have to handle views that have a `CompositeImplicitAutograd kernel` though (it just decomposes them instead). Since `_make_dual` is CompositeImplicitAutograd, I wouldn't expect a problem. Is codegen complaining?

bdhirsh replied:
oh yep, I see this PR makes that op no longer CompositeImplicitAutograd. I'll work on a fix.

bdhirsh replied:
Actually, I think the codegen can handle this case. Do you see anything failing?

soulitzer replied:
If I name the first input of `_make_dual` `primal` instead `self` I run into the following:
```
aten/src/ATen/RegisterFunctionalization_3.cpp: In function ‘at::Tensor at::functionalization::_make_dual(c10::DispatchKeySet, const at::Tensor&, const at::Tensor&, int64_t)’:
aten/src/ATen/RegisterFunctionalization_3.cpp:82:99: error: ‘self’ was not declared in this scope; did you mean ‘c10::attr::self’?
   82 |       auto out = at::functionalization::impl::create_functional_tensor_with_view_meta(tmp_output, self, view_meta);
      |                                                                                                   ^~~~
      |                                                                                                   c10::attr::self

```

In `RegisterFunctionalization_3.cpp` it looks like it is generating a function with signature `at::Tensor _make_dual(c10::DispatchKeySet dispatchKeySet, const at::Tensor & primal, const at::Tensor & tangent, int64_t level)`, but later in the function we call into `at::functionalization::impl::create_functional_tensor_with_view_meta(tmp_output, self, view_meta);` which references `self`. Perhaps we should just the name of the input instead of `self` here?

bdhirsh replied:
Ah yep, thanks! I'll make a patch.

soulitzer commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -110,6 +110,31 @@ Tensor _fw_primal(c10::DispatchKeySet ks, const Tensor & self, int64_t level) {
   return result;
 }
 
+// We need this so that set_fw_grad properly detects that _make_dual is
```
Note to self: finish this comment

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -337,7 +337,8 @@
   if (${out_arg}_new_fw_grad[i].defined()) {
   // The hardcoded 0 here will need to be updated once we support multiple levels.
-    ${out_arg}[i]._set_fw_grad(${out_arg}_new_fw_grad[i], /* level */ 0, /* is_inplace_op */ ${is_inplace});
+    ${out_arg}[i]._set_fw_grad(
```
Note to self: undo this

albanD commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -110,6 +110,31 @@ Tensor _fw_primal(c10::DispatchKeySet ks, const Tensor & self, int64_t level) {
+  auto& tangent_ = unpack(tangent, "tangent", 0);
+  std::shared_ptr<ViewBackward0> grad_fn;
+  if (compute_requires_grad(primal_)) {
+    grad_fn = std::make_shared<ViewBackward0>(); // Why is the
```
missing comment?

albanD commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -110,6 +110,31 @@ Tensor _fw_primal(c10::DispatchKeySet ks, const Tensor & self, int64_t level) {
+  }
+
+  result._set_fw_grad(tangent_, level,  /* is_inplace_op */ false);
+  TORCH_CHECK(level == 0, "Invalid level given to _make_dual");
```
nit: check before using it

albanD commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -9,10 +9,7 @@ namespace native {
-  auto dual_tensor = primal.view(primal.sizes());
-  dual_tensor._set_fw_grad(tangent, level, /* is_inplace_op */ false);
-  return dual_tensor;
+  return primal.view(primal.sizes());
```
This function can be deleted. We never go below the inplace or view kernel.

TODO: double check behavior of these functions (fw_primal and make_dual) in inference mode.

albanD commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -78,6 +78,11 @@ Tensor FunctionalInverses::_fw_primal_inverse(const at::Tensor& base, const at::
     return Tensor();
 }
 
+Tensor FunctionalInverses::_make_dual_inverse(const at::Tensor& base, const at::Tensor& mutated_view, const at::Tensor& tangent, int64_t level) {
```
Where is that used? Or is there some automatic registration system here?

soulitzer replied:
It is referenced in an auto generated function for the functionalization pass, so I had to add it to get it to compile. Maybe @bdhirsh knows more?

albanD reviewed (commented):
Looks good. Just small comments.

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).

zou3519 commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -4,16 +4,11 @@
+// because codegen assumes that a function of this name exists in the native namespace
+// TODO: check the behavior of this and _fw_primal in inference mode
+Tensor _make_dual(const Tensor& primal, const Tensor& tangnet, int64_t level) {
+  AT_ERROR("This is just a stub.");
```
nit: TORCH_CHECK? or TORCH_INTERNAL_ASSERT if this is never meant to be called

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -8,7 +8,8 @@ namespace native {
 // TODO: check the behavior of this and _fw_primal in inference mode
 Tensor _make_dual(const Tensor& primal, const Tensor& tangnet, int64_t level) {
-  AT_ERROR("This is just a stub.");
+  TORCH_INTERNAL_ASSERT("This is just a stub.");
```
I think you need a boolean: `TORCH_INTERNAL_ASSERT(false, "This is just a stub")`. Otherwise the string evaluates to True

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

albanD reviewed (approved):
Ok!

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).

soulitzer committed:
Update on "Add internal assert for tangent layout mismatch for view ops"


This PR:
 - Asserts that if the output is a view, then the `is_same_metadata` must return `true`. Otherwise, we are performing a copy. 
 - unless we are being called from `make_dual` which can allow the tangent and primal to have different layouts, because it is not forward differentiable.
 - To make this possible, we add `is_make_dual` as a parameter. ~The alternative is to make `make_dual` non-composite, and then we can rely on its `view_info` for differentiability information. This also assumes that the only composite function that calls `set_fw_grad` is `make_dual`.~
 - Batching rules now preserve storage offset for view operator that return non-zero storage offset

Differential Revision: [D31842021](https://our.internmc.facebook.com/intern/diff/D31842021)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842021).
